host_reference/domain_wall_dslash_reference.cpp

```diff
@@ -975,6 +1235,216 @@ void mdw_matpc(void *out, void **gauge, void *in, double _Complex *kappa_b, doub
+    mdw_dslash_4_pre(out, gauge, tmp, parity[1], dagger, precision, gauge_param, mferm, b5, c5, true);
+    for (int xs = 0; xs < Ls; xs++) {
+      cxpay((char *)in + precision * Vh * spinor_site_size * xs, kappa2[xs],
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else if (!symmetric && !dagger) {
+    mdw_dslash_4_pre(out, gauge, in, parity[1], dagger, precision, gauge_param, mferm, b5, c5, true);
+    dslash_4_4d(tmp, gauge, out, parity[0], dagger, precision, gauge_param, mferm);
+    mdw_eofa_m5inv(out, tmp, parity[1], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    mdw_dslash_4_pre(tmp, gauge, out, parity[0], dagger, precision, gauge_param, mferm, b5, c5, true);
+    dslash_4_4d(out, gauge, tmp, parity[1], dagger, precision, gauge_param, mferm);
+    mdw_eofa_m5(tmp, in, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    for (int xs = 0; xs < Ls; xs++) {
+      cxpay((char *)tmp + precision * Vh * spinor_site_size * xs, kappa2[xs],
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else if (!symmetric && dagger) {
+    dslash_4_4d(out, gauge, in, parity[0], dagger, precision, gauge_param, mferm);
+    mdw_dslash_4_pre(tmp, gauge, out, parity[1], dagger, precision, gauge_param, mferm, b5, c5, true);
+    mdw_eofa_m5inv(out, tmp, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    dslash_4_4d(tmp, gauge, out, parity[1], dagger, precision, gauge_param, mferm);
+    mdw_dslash_4_pre(out, gauge, tmp, parity[0], dagger, precision, gauge_param, mferm, b5, c5, true);
+    mdw_eofa_m5(tmp, in, parity[0], dagger, mferm, m5, b, c, mq1, mq2, mq3, eofa_pm, eofa_shift, precision);
+    for (int xs = 0; xs < Ls; xs++) {
+      cxpay((char *)tmp + precision * Vh * spinor_site_size * xs, kappa2[xs],
+            (char *)out + precision * Vh * spinor_site_size * xs, Vh * spinor_site_size, precision);
+    }
+  } else {
+    errorQuda("Unsupported matpc_type=%d dagger=%d", matpc_type, dagger);
+  }
+
+  free(tmp);
+}
+
+// helper for creating extended gauge fields
+static cpuGaugeField *create_extended_gauge(void **gauge, QudaGaugeParam &gauge_param, const int *R)
+{
+  GaugeFieldParam gauge_field_param(gauge, gauge_param);
+  cpuGaugeField cpu(gauge_field_param);
+
+  gauge_field_param.ghostExchange = QUDA_GHOST_EXCHANGE_EXTENDED;
+  gauge_field_param.create = QUDA_ZERO_FIELD_CREATE;
+  for (int d = 0; d < 4; d++) {
+    gauge_field_param.x[d] += 2 * R[d];
+    gauge_field_param.r[d] = R[d];
+  }
+  cpuGaugeField *padded_cpu = new cpuGaugeField(gauge_field_param);
+
+  copyExtendedGauge(*padded_cpu, cpu, QUDA_CPU_FIELD_LOCATION);
+  padded_cpu->exchangeExtendedGhost(R, true); // Do comm to fill halo = true
+
+  return padded_cpu;
+}
+
+inline int index_4d_cb_from_coordinate_4d(const int coordinate[4], const int dim[4])
```

Comment:
I won't hold up the PR on this b/c these host routines are a mess (in general, not yours specifically), but these routines are likely redundant of the routines around here: https://github.com/lattice/quda/blob/develop/utils/host_utils.h#L128

Reply:
I actually didn't find the one I wanted in the `utils/host_utils.h` ...

Reply:
Can you move it there?